Simplenote/src/main/java/com/automattic/simplenote/models/Note.java

```diff
@@ -449,6 +449,14 @@ public void setPublished(boolean isPublished) {
         }
     }
 
+    public void removeTag(String tag) {
+        List<String> tags = getTags();
+        tags.remove(tag);
```

Comment:
I'm sure you were planning on updating this as well, but we want to compare by tag hash here and not by tag name, or better, use some kind of `Tag.equals` call (which doesn't exist so probably best not to add in this PR)

```java
List<String> tags = getTags();
String tagHash = TagUtils.hashTag(tag);

for (String tagName : tags) {
    if (TagUtils.hashTag(tag) == tagHash) {
       tags.remove(tag);
    }
}

setTags(tags);
save();
```

Reply:
There are actually some methods in the `Note` model we can use to check tags:

```java
   public boolean hasTag(String tag) {
        List<String> tags = getTags();
        String tagLower = tag.toLowerCase();
        for (String tagName : tags) {
            if (tagLower.equals(tagName.toLowerCase())) return true;
        }
        return false;
    }

    public boolean hasTag(Tag tag) {
        return hasTag(tag.getSimperiumKey());
    }
```

Let me check how I can use them to check the tags.

Reply:
hm. I would recommend against using that method due to it being inaccurate - in separate work one of us should update that. also, it seems silly to iterate over all the tags inside a loop where we're already iterating over the tags.

I only meant to suggest that we rely on the `tagHash` here instead of the name. the whole `Tag.equals()` idea is just that basically. it wouldn't be more robust than…

```
// TagUtils.java or Tag.java
public static boolean areEqual(String left, String right) {
  return hashTag(left) == hashTag(right);
}
```

keeping it local and _not_ abstracting is probably most helpful at this point since `Tag`, `TagUtil`, and `Bucket` are all so jumbled and coupled together. in an ideal world I think that the `Note` object would have something like a `TagSet` property which itself held `Tag` objects and this would all be properly abstracted away from the data. for now it's just a jumble that would be hard to untangle 🤷‍♂️

Reply:
Yeah, I misread your first comment sorry (which had the solution). Fixed in a07747c09d8a98574723b9022e697c1f5a81f9e6.

Reply:
> in an ideal world I think that the Note object would have something like a TagSet property which itself held Tag objects and this would all be properly abstracted away from the data. for now it's just a jumble that would be hard to untangle

That would be interesting. It would be much simpler.